WebGLRenderer: Add mipmap & compressed texture handling to .copyTextureToTexture().

frankolivier opened the pull request:
Fixes #18736 
Fixes #18821

Commit:
Make copyTextureToTexture generate mipmaps if needed

Fixes #14423

Mugen87 commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2820,6 +2820,8 @@ function WebGLRenderer( parameters ) {
 
 		}
 
+		if (dstTexture.generateMipmaps) _gl.generateMipmap( _gl.TEXTURE_2D );
```
Instead of generating them, is it possible to copy the existing ones?

Mugen87 replied:
I'm not sure this check is safe. What if the flag is set to `false` since mipmaps were defined manually?

Commit:
Update WebGLRenderer.js

Added check for mipmap level 0

frankolivier commented:
Added a check for mipmap level 0; added 0 as the default level for this function

IMO the check is safe - if the developer wanted to manually handle mipmaps ( by setting dstTexture.generateMipmaps = `false` ) _gl.generateMipmap() should not be called in the added code.

donmccurdy commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2801,7 +2801,7 @@ function WebGLRenderer( parameters ) {
 	};
 
-	this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level ) {
+	this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level = 0 ) {
```
Perhaps we're inevitably headed toward using ES6, but for now there are no occurrences of default parameters in the library (that I know of?) and our Rollup build does not do a general-purpose ES6->ES5 transpile. Until an explicit decision is made to start using ES6, I think it'd be best to do an old-fashioned `level = level || 0` instead here.

Commit:
Update WebGLRenderer.js

Changed from default parameter to  level = level || 0

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2803,6 +2803,8 @@ function WebGLRenderer( parameters ) {
 
 	this.copyTextureToTexture = function ( position, srcTexture, dstTexture, level ) {
 
+		level = level || 0;
```
I guess this makes it more clear?
```js
if ( level === undefined ) level = 0;
```

frankolivier replied:
Will change to `if ( level === undefined ) level = 0;`

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -2812,14 +2814,16 @@ function WebGLRenderer( parameters ) {
 
 		}
 
+		if ( level == 0 && dstTexture.generateMipmaps ) _gl.generateMipmap( _gl.TEXTURE_2D );
```
I am confused now...

Is this the intention?

```js
if ( level === undefined && dstTexture.generateMipmaps ) _gl.generateMipmap( _gl.TEXTURE_2D );
```

frankolivier replied:
No... the intention is only to autogenerate mipmaps if the developer is updating the level 0 texture, and when the generateMipmaps flag is set.

mrdoob replied:
Ah! Okay.

Do you mind adding a comment?

```js
// Generate mipmaps only when copying level 0
```

frankolivier replied:
Added

Commit:
Clarified level = 0 default parameter code

Commit:
Added comment

Mugen87 commented:
Can you please remove `build/three.module.js ` from your PR?

frankolivier committed:
removing build file

Commit:
Revert "Added comment"

This reverts commit 6598637a7ee942c4098b5c952f73c03c30ac987d.

Commit:
Merge branch 'dev' of https://github.com/frankolivier/three.js into dev

Mugen87 commented:
Um, now you have completely removed `build/three.module.js`. The idea was to just undo your changes to `build/three.module.js`.

Commit:
Revert back to state before build file was added

Commit:
Redo changes

mrdoob commented:
> Fixes #14423

Hmm, do you mind explaining how this fixes #14423?

frankolivier commented:
Ah, my bad - it would not actually fix #14423. 

I've filed a new issue #18736 for the actual issue I was trying to fix for my project.

Commit:
Add compressed texture support to WebGLRenderer.copyTextureToTexture()

mrdoob commented:
Thanks!
